rsmtool/rsmpredict.py

```diff
@@ -25,6 +28,165 @@
+    h1_mean : float, optional
+       The mean of the human scores in the training set also used to re-scale
+       the predictions. May be read from the "postprocessing_params.csv" file
+       under the "output" directory of the RSMTool experiment used to train
+       the model. If ``None``, predictions will not be scaled.
+       Defaults to ``None``.
+    h1_sd : float, optional
+       The standard deviation of the human scores in the training set used to
+       re-scale the predictions. May be read from the "postprocessing_params.csv"
+       file under the "output" directory of the RSMTool experiment used to train
+       the model. If ``None``, predictions will not be scaled.
+       Defaults to ``None``.
+    logger : logging object, optional
+        A logging object. If ``None`` is passed, get logger from ``__name__``.
+        Defaults to ``None``.
+
+    Returns
+    -------
+    dict[str, float]
+        A dictionary containing the raw, scaled, trimmed, and rounded
+        predictions for the input features. It always contains the
+        "raw" key and may contain the following additional keys depending
+        on the availability of the various optional arguments:
+        "raw_trim", "raw_trim_round", "scale", "scale_trim",
+        and "scale_trim_round".
+
+    Raises
+    ------
+    ValueError
+        If ``input_features`` contains any non-numeric features.
+    """
+    # initialize a logger if none provided
+    logger = logger if logger else logging.getLogger(__name__)
+
+    # instantiate a feature preprocessor
+    preprocessor = FeaturePreprocessor(logger=logger)
+
+    # convert the given features to a data frame and add the "spkitemid" column
+    df_input_features = pd.DataFrame([input_features])
+    df_input_features["spkitemid"] = "RESPONSE"
+
+    # preprocess the input features so that they match what the model expects
+    try:
+        df_processed_features, _ = preprocessor.preprocess_new_data(
+            df_input_features, df_feature_info
+        )
+    except ValueError:
+        raise ValueError("Input features must not contain non-numeric values.") from None
+
+    # now compute the raw prediction for the given features
+    df_predictions = modeler.predict(df_processed_features)
+
+    # trim raw predictions if ``trim_min`` and ``trim_max`` were specified
+    if trim_min and trim_max:
+        df_predictions["raw_trim"] = preprocessor.trim(
+            df_predictions["raw"], trim_min, trim_max, trim_tolerance
+        )
+        df_predictions["raw_trim_round"] = np.rint(df_predictions['raw_trim']).astype("int64")
+
+    # compute scaled predictions if the scaling parameters are available
+    if train_predictions_mean and train_predictions_sd and h1_mean and h1_sd:
+        df_predictions["scale"] = (
+            (df_predictions["raw"] - train_predictions_mean) / train_predictions_sd
+        ) * h1_sd + h1_mean
+
+        # apply trimming to scaled scores too, if possible
```

Comment:
Can we combine both of the trim max and min if statements on L167 and L180 into one? Once outside the scope of the if statement on L174, we can just iterate over the columns of df_predictions?

I was thinking something like this:

```
# now compute the raw prediction for the given features
    df_predictions = modeler.predict(df_processed_features)

    # compute scaled predictions if the scaling parameters are available
    if train_predictions_mean and train_predictions_sd and h1_mean and h1_sd:
         df_predictions["scale"] = (
             (df_predictions["raw"] - train_predictions_mean) / train_predictions_sd
         ) * h1_sd + h1_mean

    # trim predictions if ``trim_min`` and ``trim_max`` were specified
    if trim_min and trim_max:
        for score_proc in df_predictions.columns:

            df_predictions[f"{score_proc}_trim"] = preprocessor.trim(
                df_predictions[score_proc], trim_min, trim_max, trim_tolerance
            )
            df_predictions[f"{score_proc}_trim_round"] = np.rint(df_predictions[f'{score_proc}_trim']).astype("int64")
```

Reply:
Neat! However, we can't just use all of `df_predictions.columns` directly since that also has `spkitemid` at this point. We'll need to filter that out but that's still more readable, thanks!